utils/retrieve-state/src/lib.rs

```diff
@@ -171,38 +176,94 @@ pub async fn download_block_with_deploys(
     })
 }
 
+/// Download block files from the highest (provided) block hash to genesis.
 pub async fn download_blocks(
     client: &mut Client,
     url: &str,
     chain_download_path: impl AsRef<Path>,
-    mut block_hash: BlockHash,
-    until_height: u64,
+    highest_block: Option<&BlockIdentifier>,
 ) -> Result<Vec<DirEntry>, anyhow::Error> {
     if !chain_download_path.as_ref().exists() {
         tokio::fs::create_dir_all(&chain_download_path).await?;
     }
-    let mut start = Instant::now();
-    loop {
-        let block_with_deploys = download_block_with_deploys(client, url, block_hash).await?;
-        block_with_deploys.save(&chain_download_path).await?;
+    let mut maybe_next_block_hash: Option<BlockHash> = match highest_block.as_ref() {
+        Some(block_identifier) => {
+            download_or_read_block_file_and_extract_parent_hash(
+                client,
+                url,
+                &chain_download_path,
+                *block_identifier,
+            )
+            .await?
+        }
+        // Get the highest block.
+        None => get_block_and_download(client, url, &chain_download_path, None).await?,
+    };
+    while let Some(next_block_hash) = maybe_next_block_hash {
+        maybe_next_block_hash = download_or_read_block_file_and_extract_parent_hash(
+            client,
+            url,
+            &chain_download_path,
+            &BlockIdentifier::Hash(next_block_hash),
+        )
+        .await?;
+    }
+    Ok(offline::get_block_files(chain_download_path))
+}
 
-        if block_with_deploys.block.header.height == until_height {
-            break;
+async fn download_or_read_block_file_and_extract_parent_hash(
+    client: &mut Client,
+    url: &str,
+    chain_download_path: &impl AsRef<Path>,
+    block_identifier: &BlockIdentifier,
+) -> Result<Option<BlockHash>, anyhow::Error> {
+    match get_block_file(&chain_download_path, block_identifier) {
+        None => {
+            Ok(
+                get_block_and_download(client, url, chain_download_path, Some(*block_identifier))
+                    .await?,
+            )
         }
-        block_hash = block_with_deploys.block.header.parent_hash;
-        if block_with_deploys.block.header.height % 1000 == 0 {
-            println!(
-                "downloaded block at height {} in {}ms",
-                block_with_deploys.block.header.height,
-                start.elapsed().as_millis()
-            );
-            start = Instant::now();
+        Some(block_file) => {
+            let block_with_deploys = read_block_file(&block_file).await?;
+            if block_with_deploys.block.header.height != 0 {
+                Ok(Some(block_with_deploys.block.header.parent_hash))
+            } else {
+                Ok(None)
+            }
         }
     }
-    println!("finished downloading blocks");
-    Ok(offline::get_block_files(chain_download_path))
 }
 
+async fn get_block_and_download(
```

Comment:
As discussed, this kind of kills my current system where I use an SD Card for storing these files.

Should we ticket using LMDB for this?

Reply:
Ticketed here: https://github.com/casper-network/casper-node/issues/2121